File: synpay-dashboard-python-api/routes/notifications.py
"""
Notification and Email Queue API routes.

These endpoints are consumed by the frontend (Next.js) and forward
all requests to the Spring Boot Integration Core for RBAC enforcement,
business logic, and audit logging.

Responsibilities:
- Forward Authorization header unchanged
- Preserve request/response format
- Propagate HTTP status codes and error messages
"""

import logging
from typing import Optional

from fastapi import APIRouter, Depends, Query, Request
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

from dependencies import extract_token, get_client_ip
from spring_client import forward_request

logger = logging.getLogger("routes.notifications")

# Two separate routers for clarity and correct path prefixes
notifications_router = APIRouter(prefix="/api/notifications", tags=["Notifications"])
email_queue_router = APIRouter(prefix="/api/email-queue", tags=["Email Queue"])


# ── Request schemas ───────────────────────────────────────────────

class NotificationEventBody(BaseModel):
    """Request body for triggering notification events (internal use)."""
    trigger_event: str = Field(..., alias="triggerEvent")
    context_data: dict | None = Field(None, alias="contextData")

    model_config = {"populate_by_name": True}


# ── Helpers ───────────────────────────────────────────────────────

def _spring_response(response) -> JSONResponse:
    """Convert an httpx response into a FastAPI JSONResponse."""
    return JSONResponse(
        status_code=response.status_code,
        content=response.json(),
    )


# ── GET /api/notifications ──────────────────────────────────────

@notifications_router.get("")
async def list_notifications(
    request: Request,
    token: str = Depends(extract_token),
    page: int = Query(0, ge=0),
    size: int = Query(20, ge=1, le=100),
    is_read: Optional[bool] = Query(None),
):
    """
    Get a paginated list of notifications for the current user.
    Optional filter by read status.

    Query parameters:
    - page: Page number (default 0)
    - size: Items per page (default 20, max 100)
    - is_read: Filter by read status (true|false|null for all)

    Forwards to: GET /internal/notifications
    """
    params: dict = {"page": page, "size": size}
    if is_read is not None:
        params["is_read"] = is_read

    response = await forward_request(
        "GET",
        "/internal/notifications",
        token,
        params=params,
        forwarded_for=get_client_ip(request),
        user_agent=request.headers.get("User-Agent"),
    )
    return _spring_response(response)


# ── PATCH /api/notifications/{id}/read ──────────────────────────

@notifications_router.patch("/{notification_id}/read")
async def mark_notification_as_read(
    notification_id: int,
    request: Request,
    token: str = Depends(extract_token),
):
    """
    Mark a specific notification as read.
    Ownership is verified by Spring Boot.

    Forwards to: PATCH /internal/notifications/{id}/read
    """
    response = await forward_request(
        "PATCH",
        f"/internal/notifications/{notification_id}/read",
        token,
        forwarded_for=get_client_ip(request),
        user_agent=request.headers.get("User-Agent"),
    )
    return _spring_response(response)


# ── GET /api/notifications/admin ─────────────────────────────────

@notifications_router.get("/admin")
async def get_admin_notifications_view(
    request: Request,
    token: str = Depends(extract_token),
    page: int = Query(0, ge=0),
    size: int = Query(20, ge=1, le=100),
):
    """
    Admin view of all user notifications (requires notification.admin permission).

    Query parameters:
    - page: Page number (default 0)
    - size: Items per page (default 20, max 100)

    Forwards to: GET /internal/notifications/admin
    """
    params: dict = {"page": page, "size": size}

    response = await forward_request(
        "GET",
        "/internal/notifications/admin",
        token,
        params=params,
        forwarded_for=get_client_ip(request),
        user_agent=request.headers.get("User-Agent"),
    )
    return _spring_response(response)


# ── GET /api/email-queue ────────────────────────────────────────

@email_queue_router.get("")
async def get_email_queue(
    request: Request,
    token: str = Depends(extract_token),
    status: str = Query(..., description="Filter by status: PENDING, SENT, or FAILED"),
    page: int = Query(0, ge=0),
    size: int = Query(20, ge=1, le=100),
):
    """
    Get paginated email queue entries filtered by status.
    Requires email_queue.view permission (admin only).

    Query parameters:
    - status: Email status filter (PENDING|SENT|FAILED) - REQUIRED
    - page: Page number (default 0)
    - size: Items per page (default 20, max 100)

    Forwards to: GET /internal/email-queue
    """
    params: dict = {"status": status, "page": page, "size": size}

    response = await forward_request(
        "GET",
        "/internal/email-queue",
        token,
        params=params,
        forwarded_for=get_client_ip(request),
        user_agent=request.headers.get("User-Agent"),
    )
    return _spring_response(response)


# ── POST /api/email-queue/{id}/retry ────────────────────────────

@email_queue_router.post("/{email_id}/retry")
async def retry_failed_email(
    email_id: int,
    request: Request,
    token: str = Depends(extract_token),
):
    """
    Retry a failed email by moving it back to PENDING status.
    Only applies to emails with FAILED status.
    Requires email_queue.view permission (admin only).

    Forwards to: POST /internal/email-queue/{id}/retry
    """
    response = await forward_request(
        "POST",
        f"/internal/email-queue/{email_id}/retry",
        token,
        forwarded_for=get_client_ip(request),
        user_agent=request.headers.get("User-Agent"),
    )
    return _spring_response(response)


# Export both routers for main.py to include
router = notifications_router  # For backward compatibility with import in main.py

File: synpay-dashboard-frontend/api/notifications.ts
/**
 * Notification & Email Queue API Client
 *
 * Calls the FastAPI gateway at /api/notifications, which forwards
 * all requests to the Spring Boot Integration Core.
 *
 * RBAC is enforced server-side — this layer only handles HTTP transport
 * and type-safe response mapping.
 */

import { authHeader } from '@/lib/auth'

const API_BASE = process.env.NEXT_PUBLIC_API_URL ?? 'http://localhost:8000'

// ── API Envelope ─────────────────────────────────────────────────

export interface ApiEnvelope<T> {
  success: boolean
  message?: string
  data?: T
}

// ── Notification Response Types (mirror Spring Boot DTOs) ────────

/** Single notification item */
export interface NotificationItem {
  notification_id: number
  account_id: number
  template_code: string
  title: string
  message: string
  is_read: boolean
  related_resource?: string | null
  related_resource_id?: string | null
  created_at: string
}

/** Paginated notifications response */
export interface NotificationPage {
  content: NotificationItem[]
  page: number
  size: number
  totalElements: number
  totalPages: number
  unreadCount: number
}

/** Single email queue entry */
export interface EmailQueueItem {
  email_id: number
  to_email: string
  subject: string
  status: 'PENDING' | 'SENT' | 'FAILED'
  retry_count: number
  last_error?: string | null
  created_at: string
  sent_at?: string | null
}

/** Paginated email queue response */
export interface EmailQueuePage {
  content: EmailQueueItem[]
  page: number
  size: number
  totalElements: number
  totalPages: number
}

// ── Query Filters ────────────────────────────────────────────────

export interface NotificationListParams {
  page?: number
  size?: number
  is_read?: boolean // undefined = all, true = read, false = unread
}

export interface EmailQueueListParams {
  status: 'PENDING' | 'SENT' | 'FAILED'
  page?: number
  size?: number
}

// ── Helpers ──────────────────────────────────────────────────────

async function get<T>(path: string, params?: Record<string, any>): Promise<T> {
  const url = new URL(`${API_BASE}${path}`)
  if (params) {
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined && value !== null) {
        url.searchParams.append(key, String(value))
      }
    })
  }

  const res = await fetch(url.toString(), {
    method: 'GET',
    headers: { 'Content-Type': 'application/json', ...authHeader() },
  })
  if (!res.ok) {
    const body = await res.json().catch(() => null)
    throw new Error(body?.message ?? `Notification API error: ${res.status}`)
  }
  const envelope: ApiEnvelope<T> = await res.json()
  if (!envelope.success || !envelope.data) {
    throw new Error(envelope.message ?? 'Failed to fetch notification data')
  }
  return envelope.data
}

async function patch<T>(path: string): Promise<T> {
  const url = `${API_BASE}${path}`
  const res = await fetch(url, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json', ...authHeader() },
  })
  if (!res.ok) {
    const body = await res.json().catch(() => null)
    throw new Error(body?.message ?? `Notification API error: ${res.status}`)
  }
  const envelope: ApiEnvelope<T> = await res.json()
  if (!envelope.success || !envelope.data) {
    throw new Error(envelope.message ?? 'Failed to update notification')
  }
  return envelope.data
}

async function post<T>(path: string): Promise<T> {
  const url = `${API_BASE}${path}`
  const res = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...authHeader() },
  })
  if (!res.ok) {
    const body = await res.json().catch(() => null)
    throw new Error(body?.message ?? `Notification API error: ${res.status}`)
  }
  const envelope: ApiEnvelope<T> = await res.json()
  if (!envelope.success || !envelope.data) {
    throw new Error(envelope.message ?? 'Failed to perform action')
  }
  return envelope.data
}

// ── Notification API Functions ───────────────────────────────────

/**
 * Get paginated list of notifications for current user.
 * Optional filter by read status.
 * Requires: notification.read
 */
export function fetchNotifications(
  params?: NotificationListParams,
): Promise<NotificationPage> {
  const queryParams: Record<string, any> = {
    page: params?.page ?? 0,
    size: params?.size ?? 20,
  }
  if (params?.is_read !== undefined) {
    queryParams.is_read = params.is_read
  }
  return get<NotificationPage>('/api/notifications', queryParams)
}

/**
 * Mark a specific notification as read.
 * Current user must own the notification.
 * Requires: notification.mark_read
 */
export function markNotificationAsRead(
  notificationId: number,
): Promise<NotificationItem> {
  return patch<NotificationItem>(`/api/notifications/${notificationId}/read`)
}

/**
 * Get admin view of all notifications.
 * Requires: notification.admin
 */
export function fetchAdminNotifications(
  params?: { page?: number; size?: number },
): Promise<NotificationPage> {
  const queryParams: Record<string, any> = {
    page: params?.page ?? 0,
    size: params?.size ?? 20,
  }
  return get<NotificationPage>('/api/notifications/admin', queryParams)
}

// ── Email Queue API Functions ────────────────────────────────────

/**
 * Get paginated email queue entries filtered by status.
 * Requires: email_queue.view (admin only)
 */
export function fetchEmailQueue(
  params: EmailQueueListParams,
): Promise<EmailQueuePage> {
  const queryParams: Record<string, any> = {
    status: params.status,
    page: params.page ?? 0,
    size: params.size ?? 20,
  }
  return get<EmailQueuePage>('/api/email-queue', queryParams)
}

/**
 * Retry a failed email (move back to PENDING status).
 * Only applies to FAILED emails.
 * Requires: email_queue.view (admin only)
 */
export function retryFailedEmail(emailId: number): Promise<EmailQueueItem> {
  return post<EmailQueueItem>(`/api/email-queue/${emailId}/retry`)
}

// ── Notification Service Helper ──────────────────────────────────

/**
 * Get unread notification count.
 * Convenience helper that fetches all unread notifications.
 */
export async function getUnreadNotificationCount(): Promise<number> {
  const page = await fetchNotifications({ is_read: false, size: 1 })
  return page.unreadCount
}
